src/backend/distributed/utils/placement_connection.c

```diff
@@ -0,0 +1,483 @@
+
+	/* list of connections to remote nodes */
+	dlist_head connectionReferences;
+
+	/* membership in ConnectionShardHashEntry->placementConnections */
+	dlist_node shardNode;
+} ConnectionPlacementHashEntry;
+
+/* hash table */
+static HTAB *ConnectionPlacementHash;
+
+
+/*
+ * Hash table mapping shard ids to placements.
+ *
+ * This is used to track whether placements of a shard have to be marked
+ * invalid after a failure, or whether a coordinated transaction has to be
+ * aborted, to avoid all placements of a shard to be marked invalid.
+ */
+
+/* hash key */
+typedef struct ConnectionShardHashKey
+{
+	uint64 shardId;
+} ConnectionShardHashKey;
+
+/* hash entry */
+typedef struct ConnectionShardHashEntry
+{
+	ConnectionShardHashKey key;
+	dlist_head placementConnections;
+} ConnectionShardHashEntry;
+
+/* hash table itself */
+static HTAB *ConnectionShardHash;
+
+
+static void AssociatePlacementWithShard(ConnectionPlacementHashEntry *placementEntry,
+										ShardPlacement *placement);
+static bool CheckShardPlacements(ConnectionShardHashEntry *shardEntry, bool preCommit,
+								 bool using2PC);
+
+
+/*
+ * GetPlacementConnection establishes a connection for a placement.
+ *
+ * See StartPlacementConnection for details.
+ */
+MultiConnection *
+GetPlacementConnection(uint32 flags, ShardPlacement *placement)
+{
+	MultiConnection *connection = StartPlacementConnection(flags, placement);
+
+	FinishConnectionEstablishment(connection);
+	return connection;
+}
+
+
+/*
+ * StartPlacementConnection() initiates a connection to a remote node,
+ * associated with the placement and transaction.
+ *
+ * The connection is established as the current user & database.
+ *
+ * See StartNodeUserDatabaseConnection for details.
+ *
+ * Flags have the corresponding meaning from StartNodeUserDatabaseConnection,
+ * except that two additional flags have an effect:
+ * - FOR_DML - signal that connection is going to be used for DML (modifications)
+ * - FOR_DDL - signal that connection is going to be used for DDL
+ *
+ * Only one connection associated with the placement may have FOR_DML or
+ * FOR_DDL set. This restriction prevents deadlocks and wrong results due to
+ * in-progress transactions.
+ */
+MultiConnection *
+StartPlacementConnection(uint32 flags, ShardPlacement *placement)
+{
+	ConnectionPlacementHashKey key;
+	ConnectionPlacementHashEntry *placementEntry = NULL;
+	bool found = false;
+	ConnectionReference *returnConnectionReference = NULL;
+	dlist_iter it;
+
+	key.placementId = placement->placementId;
+
+	/* lookup relevant hash entry */
+	placementEntry = hash_search(ConnectionPlacementHash, &key, HASH_ENTER, &found);
+	if (!found)
+	{
+		dlist_init(&placementEntry->connectionReferences);
+		placementEntry->failed = false;
+	}
+
+
+	/*
+	 * Check whether any of the connections already associated with the
+	 * placement can be reused, or violates FOR_DML/FOR_DDL constraints.
+	 */
+	dlist_foreach(it, &placementEntry->connectionReferences)
```

Comment:
I'm unconvinced. I mean I have no problem adding a modifyingConnection, but I don't think the above coding makes things simpler. Especially as we also have to deal with the userName matching.